src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,532 @@
+        {
+            var symbol = s_semanticModel.GetSymbolInfo(variableDeclaration.Type).Symbol as ITypeSymbol;
+            isBuilders = symbol.IsBuilderDefinition();
+        }
+        else if (syntaxNode is AssignmentExpressionSyntax assignmentExpression)
+        {
+            var variableIdentifier = assignmentExpression.Left.DescendantNodesAndSelf().OfType<IdentifierNameSyntax>().FirstOrDefault();
+            var symbol = s_semanticModel.GetTypeInfo(variableIdentifier).Type;
+            isBuilders = symbol.IsBuilderDefinition();
+        }
+
+        return isBuilders;
+    }
+
+    private static bool IsBlock(SyntaxNode syntaxNode) =>
+        syntaxNode switch
+        {
+            SwitchStatementSyntax or
+            SwitchSectionSyntax or
+            SwitchExpressionSyntax or
+            SwitchExpressionArmSyntax or
+            IfStatementSyntax or
+            ElseClauseSyntax or
+            ForStatementSyntax or
+            CommonForEachStatementSyntax or
+            DoStatementSyntax or
+            WhileStatementSyntax or
+            TryStatementSyntax or
+            CatchClauseSyntax or
+            FinallyClauseSyntax or
+            BlockSyntax or
+            CheckedStatementSyntax or
+            FixedStatementSyntax or
+            LabeledStatementSyntax or
+            UnsafeStatementSyntax => true,
+            _ => false
+        };
+
+    private static bool IsConditionalStatementOrClause(SyntaxNode syntaxNode) =>
+        syntaxNode switch
+        {
+            SwitchStatementSyntax or
+            SwitchSectionSyntax or
+            SwitchExpressionSyntax or
+            SwitchExpressionArmSyntax or
+            IfStatementSyntax or
+            ElseClauseSyntax or
+            ForStatementSyntax or
+            CommonForEachStatementSyntax or
+            DoStatementSyntax or
+            WhileStatementSyntax or
+            TryStatementSyntax or
+            CatchClauseSyntax => true,
+            _ => false
+        };
+
+    private static bool IsMethod(SyntaxNode syntaxNode) =>
+        syntaxNode switch
+        {
+            MethodDeclarationSyntax => true,
+            _ => false
+        };
+
+    private static bool IsConditionallyExecuted(SyntaxNode syntaxNode) =>
+        syntaxNode switch
+        {
+            CheckedStatementSyntax or
+            FinallyClauseSyntax or
+            FixedStatementSyntax or
+            UnsafeStatementSyntax or
+            LabeledStatementSyntax => false,
+            _ => IsConditionalStatementOrClause(syntaxNode.Parent) || IsMethod(syntaxNode.Parent)
+        };
+
+    private static HashSet<string> ProcessBlock(SyntaxNode syntaxNode)
+    {
+        var dirtyVariables = new HashSet<string>();
+        var isConditionalStatementOrClause = IsConditionalStatementOrClause(syntaxNode);
+        var isMethod = IsMethod(syntaxNode);
+
+        foreach (var childNode in syntaxNode.ChildNodes())
```

Comment:
Try using syntaxNode.DescendantNodes, and limit the recursion only to conditional blocks.